lib/sqlalchemy/engine/url.py

```diff
@@ -560,6 +560,20 @@ def __str__(self):
     def __repr__(self):
         return self.render_as_string()
 
+    def __copy__(self):
+        return __class__.create(
+            self.drivername,
+            self.username,
+            self.password,
+            self.host,
+            self.port,
+            self.database,
+            self.query,
+        )
+
+    def __deepcopy__(self, memo):
+        return self.__copy__()
```

Comment:
self.query is a dict so we likely need to deepcopy that also

Reply:
though if it's a dict of str->str in all cases, then not.  let me check that

Reply:
Ooops!

It still needs to be copied, though, you don't want the same mutable dict in two entries.

Reply:
it's fine the way it is, the dict is immutable (special SQLAlhcemy class) and the contents are all strings / tuples of strings / tuples .  fully immutable contents as well. thanks !

Reply:
Immutable dicts, great idea!

Thanks again for an excellent package, and I hope the rest of your weekend is fantastic.